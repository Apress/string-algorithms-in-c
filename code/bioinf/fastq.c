
#include "fastq.h"

#include <stdlib.h>
#include <string.h>


void init_fastq_iter(
    struct fastq_iter *iter,
    FILE *file
) {
    iter->file = file;
    iter->buffer = malloc(MAX_STRING_LEN + 1);
}

bool next_fastq_record(
    struct fastq_iter *iter,
    struct fastq_record *record
) {
    FILE *file = iter->file;
    char *buffer = iter->buffer;
    if (fgets(buffer, MAX_STRING_LEN, file)) {
        strcpy((char*)record->name, strtok(buffer + 1, "\n"));
        
        fgets(buffer, MAX_STRING_LEN, file);
        strcpy((char*)record->sequence, strtok(buffer, "\n"));
        
        fgets(buffer, MAX_STRING_LEN, file);
        fgets(buffer, MAX_STRING_LEN, file);
        strcpy((char*)record->quality, strtok(buffer, "\n"));
        return true;
    }
    return false;
}

void dealloc_fastq_iter(
    struct fastq_iter *iter
) {
    free(iter->buffer);
}
