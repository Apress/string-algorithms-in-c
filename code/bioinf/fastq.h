
#ifndef FASTQ_H
#define FASTQ_H

#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>

#define MAX_STRING_LEN 2048

struct fastq_iter {
    FILE *file;
    char *buffer;
};
struct fastq_record {
    char name[MAX_STRING_LEN];
    uint8_t sequence[MAX_STRING_LEN];
    char quality[MAX_STRING_LEN];
};

// FIXME: a way to report errors if we have id:0
// a malformed fastq file.
void init_fastq_iter(
    struct fastq_iter *iter,
    FILE *file
);
bool next_fastq_record(
    struct fastq_iter *iter,
    struct fastq_record *record
);
void dealloc_fastq_iter(
    struct fastq_iter *iter
);


#endif
