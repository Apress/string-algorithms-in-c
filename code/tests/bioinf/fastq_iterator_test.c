
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#include "fastq.h"


int main(int argc, char *argv[])
{
    FILE *input = fopen("test-data/test.fq", "r");
    assert(input);
    FILE *output = fopen("test-data/fastq_iterator_test_current.txt", "w");
    assert(output);
    
    struct fastq_iter iter;
    struct fastq_record record;
    init_fastq_iter(&iter, input);
    while (next_fastq_record(&iter, &record)) {
        fprintf(output, "@%s\n", record.name);
        fprintf(output, "%s\n", record.sequence);
        fprintf(output, "+\n");
        fprintf(output, "%s\n", record.quality);
    }
    dealloc_fastq_iter(&iter);
    fclose(input);
    fclose(output);

    return EXIT_SUCCESS;
}
