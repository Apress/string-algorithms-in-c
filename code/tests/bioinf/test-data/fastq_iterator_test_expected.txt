@read1
ACCTACAGACTACCATGTATCTCCATTTACCTAGTCTAG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@read2
ACCTACAGACTACCATGTATCTCCATTTACCTAGTCTAG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@read3
ACCTACAGACTACCATGTATCTCCATTTACCTAGTCTAG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@read4
ACCTACAGACTACCATGTATCTCCATTTACCTAGTCTAG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@read5
CTAGCTTTCCACACGCTGTGTGTCACTAGTGTGACTACG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@read6
ACCTACAGACTACCATGTATCTCCATTTACCTAGTCTAG
+
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
